File: campus-navigate-app/flutter_app/lib/domain/edge_editor/logic/route_metric_model.dart
import 'package:indoor_navigation/domain/common/geometry_math.dart';
import 'package:indoor_navigation/domain/edge_editor/edge_editor_models.dart';
import 'package:indoor_navigation/domain/tiled/tiled_models.dart';

RouteMetricModel createRouteMetricModel(
  List<OverlayRouteNode> routePath,
  List<RouteGraphEdgeExport> edges,
) {
  final segments = <RouteMetricSegment>[];
  var startDistanceMeters = 0.0;
  var startDistancePixels = 0.0;

  for (var index = 0; index < routePath.length - 1; index += 1) {
    final from = routePath[index];
    final to = routePath[index + 1];
    final edge = _findEdgeBetweenNodes(edges, from.nodeId, to.nodeId);
    if (edge == null) {
      throw StateError(
        'Missing EDGE distance for ${from.nodeId} -> ${to.nodeId}.',
      );
    }

    final pixelLength = distanceBetweenPoints(from, to);
    final meterLength = edge.distance;
    if (meterLength <= 0) {
      throw StateError('EDGE distance must be positive for ${edge.id}.');
    }

    final endDistanceMeters = startDistanceMeters + meterLength;
    final endDistancePixels = startDistancePixels + pixelLength;
    segments.add(
      RouteMetricSegment(
        endDistanceMeters: endDistanceMeters,
        endDistancePixels: endDistancePixels,
        fromNodeId: from.nodeId,
        meterLength: meterLength,
        pixelLength: pixelLength,
        pixelsPerMeter: pixelLength / meterLength,
        startDistanceMeters: startDistanceMeters,
        startDistancePixels: startDistancePixels,
        toNodeId: to.nodeId,
      ),
    );
    startDistanceMeters = endDistanceMeters;
    startDistancePixels = endDistancePixels;
  }

  return RouteMetricModel(
    averagePixelsPerMeter: startDistanceMeters > 0
        ? startDistancePixels / startDistanceMeters
        : 1,
    segments: segments,
    totalMeters: startDistanceMeters,
    totalPixels: startDistancePixels,
  );
}

double findPixelsPerMeterAtRoutePosition({
  required RouteMetricModel metrics,
  required RoutePosition position,
}) {
  for (final segment in metrics.segments) {
    if (position.distanceAlongRoute >= segment.startDistancePixels &&
        position.distanceAlongRoute <= segment.endDistancePixels) {
      return segment.pixelsPerMeter;
    }
  }
  return metrics.averagePixelsPerMeter;
}

double convertMetersToPixelsAtRoutePosition({
  required double meters,
  required RouteMetricModel metrics,
  required RoutePosition position,
}) {
  return meters *
      findPixelsPerMeterAtRoutePosition(metrics: metrics, position: position);
}

RouteGraphEdgeExport? _findEdgeBetweenNodes(
  List<RouteGraphEdgeExport> edges,
  String fromNodeId,
  String toNodeId,
) {
  for (final edge in edges) {
    if ((edge.from == fromNodeId && edge.to == toNodeId) ||
        (edge.from == toNodeId && edge.to == fromNodeId)) {
      return edge;
    }
  }
  return null;
}

File: tarumt-nav-app/expo-app/src/edgeEditor/routeMetricModel.ts
import { distanceBetweenPoints } from '../tiled/model/geometryMath';
import type { OverlayRouteNode, RoutePosition } from '../tiled/type';
import type { RouteGraphEdgeExport } from './routeGraphEdgeModel';

export type RouteMetricSegment = {
  endDistanceMeters: number;
  endDistancePixels: number;
  fromNodeId: string;
  meterLength: number;
  pixelLength: number;
  pixelsPerMeter: number;
  startDistanceMeters: number;
  startDistancePixels: number;
  toNodeId: string;
};

export type RouteMetricModel = {
  averagePixelsPerMeter: number;
  segments: RouteMetricSegment[];
  totalMeters: number;
  totalPixels: number;
};

export function createRouteMetricModel(
  routePath: readonly OverlayRouteNode[],
  edges: readonly RouteGraphEdgeExport[],
): RouteMetricModel {
  const segments: RouteMetricSegment[] = [];
  let startDistanceMeters = 0;
  let startDistancePixels = 0;

  for (let index = 0; index < routePath.length - 1; index += 1) {
    const from = routePath[index];
    const to = routePath[index + 1];
    if (!from || !to) {
      continue;
    }

    const edge = findEdgeBetweenNodes(edges, from.nodeId, to.nodeId);
    if (!edge) {
      throw new Error(`Missing EDGE distance for ${from.nodeId} -> ${to.nodeId}.`);
    }

    const pixelLength = distanceBetweenPoints(from, to);
    const meterLength = edge.distance;
    if (meterLength <= 0) {
      throw new Error(`EDGE distance must be positive for ${edge.id}.`);
    }

    const endDistanceMeters = startDistanceMeters + meterLength;
    const endDistancePixels = startDistancePixels + pixelLength;
    segments.push({
      endDistanceMeters,
      endDistancePixels,
      fromNodeId: from.nodeId,
      meterLength,
      pixelLength,
      pixelsPerMeter: pixelLength / meterLength,
      startDistanceMeters,
      startDistancePixels,
      toNodeId: to.nodeId,
    });
    startDistanceMeters = endDistanceMeters;
    startDistancePixels = endDistancePixels;
  }

  return {
    averagePixelsPerMeter:
      startDistanceMeters > 0 ? startDistancePixels / startDistanceMeters : 1,
    segments,
    totalMeters: startDistanceMeters,
    totalPixels: startDistancePixels,
  };
}

export function findPixelsPerMeterAtRoutePosition(input: {
  metrics: RouteMetricModel;
  position: RoutePosition;
}) {
  const segment = input.metrics.segments.find(
    (currentSegment) =>
      input.position.distanceAlongRoute >= currentSegment.startDistancePixels &&
      input.position.distanceAlongRoute <= currentSegment.endDistancePixels,
  );

  return segment?.pixelsPerMeter ?? input.metrics.averagePixelsPerMeter;
}

export function convertMetersToPixelsAtRoutePosition(input: {
  meters: number;
  metrics: RouteMetricModel;
  position: RoutePosition;
}) {
  return (
    input.meters *
    findPixelsPerMeterAtRoutePosition({
      metrics: input.metrics,
      position: input.position,
    })
  );
}

function findEdgeBetweenNodes(
  edges: readonly RouteGraphEdgeExport[],
  fromNodeId: string,
  toNodeId: string,
) {
  return edges.find(
    (edge) =>
      (edge.from === fromNodeId && edge.to === toNodeId) ||
      (edge.from === toNodeId && edge.to === fromNodeId),
  );
}
